src/openrct2/world/Map.h

```diff
@@ -198,6 +197,22 @@ void map_invalidate_selection_rect();
 bool MapCheckCapacityAndReorganise(const CoordsXY& loc, size_t numElements = 1);
 TileElement* tile_element_insert(const CoordsXYZ& loc, int32_t occupiedQuadrants, TileElementType type);
 
+template<typename T = TileElement> T* MapGetFirstTileElementWithBaseHeightBetween(const TileCoordsXYRangedZ& loc)
+{
+    TileElement* tileElement = map_get_first_element_at(loc.ToCoordsXY());
+    if (tileElement == nullptr)
+        return nullptr;
+    do
+    {
+        if (tileElement->GetType() != T::ElementType)
+            continue;
+        if (tileElement->base_height >= loc.baseZ && tileElement->base_height <= loc.clearanceZ)
+            return tileElement->as<T>();
+    } while (!(tileElement++)->IsLastForTile() && tileElement->base_height <= loc.clearanceZ);
+
+    return nullptr;
+}
```

Comment:
Try to keep as much code as you can in the .cpp file. For this you can have a non-template version in the .cpp file, then just have the template version call the non-template version and do `->as<T>` at the end.

Reply:
Thanks, that's much better indeed.